File: backend/python/app/core/limit_price.py
"""涨跌停价自算（纯函数，便于单测）。

A 股涨跌幅制度（注册制后）：
- 主板（沪 60*/深 000*/001*/002*/003*）       ±10%
- 创业板（300*/301*）、科创板（688*/689*）      ±20%
- 北交所（8**/4**/920*）                        ±30%
- 主板 ST                                       ±5%
- 创业板/科创板 ST                               仍 ±20%（注册制后与非 ST 相同）
- 北交所 ST                                      仍 ±30%
- 新股上市首日 / 科创创业前 5 日不设限            → 返回 (None, None)
"""
from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal


def board_of(code: str) -> str:
    """按代码段判定板块。兼容 baostock 的 'sh.600000' 与纯代码 '600000'。"""
    c = code.split(".")[-1]
    if c.startswith(("688", "689")):
        return "STAR"  # 科创板
    if c.startswith(("300", "301")):
        return "GEM"   # 创业板
    if c.startswith("920") or c.startswith(("8", "4")):
        return "BSE"   # 北交所
    return "MAIN"      # 主板


def limit_pct(code: str, is_st: bool) -> Decimal:
    """返回该标的当日涨跌幅比例（单边）。"""
    board = board_of(code)
    if board == "BSE":
        return Decimal("0.30")
    if board in ("GEM", "STAR"):
        return Decimal("0.20")  # 创业/科创：ST 也按 20%
    return Decimal("0.05") if is_st else Decimal("0.10")  # 主板：ST 5%，否则 10%


def compute_limit_prices(
    code: str,
    pre_close: Decimal | float | str | None,
    is_st: bool = False,
    is_first_day: bool = False,
) -> tuple[Decimal | None, Decimal | None]:
    """计算涨跌停价（四舍五入到分）；不设限 / 无有效昨收时返回 (None, None)。"""
    if is_first_day or pre_close is None:
        return None, None
    pc = Decimal(str(pre_close))
    if pc <= 0:
        return None, None
    pct = limit_pct(code, is_st)
    up = (pc * (Decimal("1") + pct)).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    down = (pc * (Decimal("1") - pct)).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
    return up, down

File: backend/go/internal/service/limit_price.go
package service

import "github.com/shopspring/decimal"

// 涨跌停价自算（与 Python core/limit_price.py 口径一致），用于 gotdx 增量采集时补全
// limit_up/limit_down（gotdx 行情不直接给涨跌停价，依板块/ST 规则现算）。
//
// A 股涨跌幅制度（注册制后）：
//   - 主板（沪 60*/深 000*/001*/002*/003*）  ±10%，ST ±5%
//   - 创业板（300*/301*）、科创板（688*/689*） ±20%（ST 同 20%）
//   - 北交所（8**/4**/920*）                  ±30%（ST 同 30%）
//   - 新股上市首日 / 不设限                    → (zero, zero, false)

const (
	boardMain = "MAIN"
	boardGEM  = "GEM"
	boardSTAR = "STAR"
	boardBSE  = "BSE"
)

// boardOf 按代码段判定板块（兼容 "sh.600000" 与纯代码 "600000"）。
func boardOf(code string) string {
	c := code
	if i := lastDotIndex(c); i >= 0 {
		c = c[i+1:]
	}
	switch {
	case hasPrefix(c, "688") || hasPrefix(c, "689"):
		return boardSTAR
	case hasPrefix(c, "300") || hasPrefix(c, "301"):
		return boardGEM
	case hasPrefix(c, "920") || hasPrefix(c, "8") || hasPrefix(c, "4"):
		return boardBSE
	default:
		return boardMain
	}
}

// limitPct 返回该标的单边涨跌幅比例。
func limitPct(code string, isST bool) decimal.Decimal {
	switch boardOf(code) {
	case boardBSE:
		return decimal.NewFromFloat(0.30)
	case boardGEM, boardSTAR:
		return decimal.NewFromFloat(0.20) // 创业/科创：ST 也按 20%
	default:
		if isST {
			return decimal.NewFromFloat(0.05)
		}
		return decimal.NewFromFloat(0.10)
	}
}

// computeLimitPrices 计算涨跌停价（四舍五入到分）；不设限 / 无有效昨收时返回 (0, 0, false)。
func computeLimitPrices(code string, preClose decimal.Decimal, isST, isFirstDay bool) (up, down decimal.Decimal, ok bool) {
	if isFirstDay || preClose.LessThanOrEqual(decimal.Zero) {
		return decimal.Zero, decimal.Zero, false
	}
	pct := limitPct(code, isST)
	one := decimal.NewFromInt(1)
	up = preClose.Mul(one.Add(pct)).Round(2)
	down = preClose.Mul(one.Sub(pct)).Round(2)
	return up, down, true
}

func lastDotIndex(s string) int {
	for i := len(s) - 1; i >= 0; i-- {
		if s[i] == '.' {
			return i
		}
	}
	return -1
}

func hasPrefix(s, prefix string) bool {
	return len(s) >= len(prefix) && s[:len(prefix)] == prefix
}
